framework/src/mesh/MooseMesh.C

```diff
@@ -600,51 +600,51 @@ MooseMesh::buildNodeList()
 {
   freeBndNodes();
 
-  /// Boundary node list (node ids and corresponding side-set ids, arrays always have the same length)
-  std::vector<dof_id_type> nodes;
-  std::vector<boundary_id_type> ids;
-  getMesh().get_boundary_info().build_node_list(nodes, ids);
-
-  int n = nodes.size();
-  _bnd_nodes.resize(n);
-  for (int i = 0; i < n; i++)
+  auto bc_tuples = getMesh().get_boundary_info().build_node_list();
+
+  int n = bc_tuples.size();
+  _bnd_nodes.clear();
+  _bnd_nodes.reserve(n);
+  for (const auto & t : bc_tuples)
   {
-    _bnd_nodes[i] = new BndNode(getMesh().node_ptr(nodes[i]), ids[i]);
-    _node_set_nodes[ids[i]].push_back(nodes[i]);
-    _bnd_node_ids[ids[i]].insert(nodes[i]);
+    auto node_id = std::get<0>(t);
+    auto bc_id = std::get<1>(t);
+
+    _bnd_nodes.push_back(new BndNode(getMesh().node_ptr(node_id), bc_id));
+    _node_set_nodes[bc_id].push_back(node_id);
+    _bnd_node_ids[bc_id].insert(node_id);
   }
 
   _bnd_nodes.reserve(_bnd_nodes.size() + _extra_bnd_nodes.size());
   for (unsigned int i = 0; i < _extra_bnd_nodes.size(); i++)
   {
     BndNode * bnode = new BndNode(_extra_bnd_nodes[i]._node, _extra_bnd_nodes[i]._bnd_id);
     _bnd_nodes.push_back(bnode);
-    _bnd_node_ids[ids[i]].insert(_extra_bnd_nodes[i]._node->id());
+    _bnd_node_ids[std::get<1>(bc_tuples[i])].insert(_extra_bnd_nodes[i]._node->id());
   }
 
-  BndNodeCompare mein_kompfare;
-
   // This sort is here so that boundary conditions are always applied in the same order
-  std::sort(_bnd_nodes.begin(), _bnd_nodes.end(), mein_kompfare);
+  std::sort(_bnd_nodes.begin(), _bnd_nodes.end(), BndNodeCompare());
 }
 
 void
 MooseMesh::buildBndElemList()
 {
   freeBndElems();
 
-  /// Boundary node list (node ids and corresponding side-set ids, arrays always have the same length)
-  std::vector<dof_id_type> elems;
-  std::vector<unsigned short int> sides;
-  std::vector<boundary_id_type> ids;
-  getMesh().get_boundary_info().build_active_side_list(elems, sides, ids);
+  auto bc_tuples = getMesh().get_boundary_info().build_active_side_list();
 
-  int n = elems.size();
-  _bnd_elems.resize(n);
-  for (int i = 0; i < n; i++)
+  int n = bc_tuples.size();
+  _bnd_elems.clear();
+  _bnd_elems.reserve(n);
+  for (const auto & t : bc_tuples)
   {
-    _bnd_elems[i] = new BndElement(getMesh().elem_ptr(elems[i]), sides[i], ids[i]);
-    _bnd_elem_ids[ids[i]].insert(elems[i]);
+    auto elem_id = std::get<0>(t);
+    auto side_id = std::get<1>(t);
+    auto bc_id = std::get<2>(t);
```

Comment:
I was going to suggest using `std::tie`, but I don't think I like that syntax as much. It looks like there's no way to use `auto` with it. I think this is probably better.

Reply:
Agreed, I think `std::tie` will essentially be obsoleted by the C++17 structured bindings.